cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -556,6 +557,10 @@ func NewControllerDescriptors() map[string]*ControllerDescriptor {
 	register(newResourceClaimControllerDescriptor())
 	register(newLegacyServiceAccountTokenCleanerControllerDescriptor())
 	register(newValidatingAdmissionPolicyStatusControllerDescriptor())
+	if utilfeature.DefaultFeatureGate.Enabled(features.SeparateTaintEvictionController) {
```

Comment:
@atiratree not quite, this controller is on by default, since it's beta, so this check ensures with the feature gate off it's not turned on. Not sure how to handle that with your controller initiation changes. So I think we can leave it as is, for now, and fix in a followup.

Reply:
this check has been moved to 

https://github.com/kubernetes/kubernetes/blob/b84e4a1333fea94218a9ee85966d464cc7d61266/cmd/kube-controller-manager/app/controllermanager.go#L683-L688

and it should work properly even when it is beta and on by default and then turned off by the user.

So it is redundant to have it in this place as well. But I don't mind merging this PR first and fixing this later.

fyi, I am trying to enforce this in https://github.com/kubernetes/kubernetes/pull/121611

Reply:
`requiredFeatureGates` field is meant as a replacement of this logic and the other gated controllers were ported to it without any issue. It is possible to gate a controller with multiple feature gates as you can see for example in `StorageVersionGarbageCollectorController`. 

I made the simplest implementation I could. The implementation can be changed in the future if more custom decision logic (not just `AND`) is required. Since this controller has just one, it should be okay in this situation as well.

The reason behind this is simple, we want to reason about controllers without running them and have the gating information available declaratively. For example we want to show available controllers in `kube-controller-manager --help`. We can then also show the gating information there.

Reply:
I kind of warned this was going to happen https://github.com/kubernetes/kubernetes/pull/120371#discussion_r1329313427 🙃

Reply:
I have updated the https://github.com/kubernetes/kubernetes/pull/121611 PR and included a test to show that the gating works properly with this new controller. Please let me know which part of the test you think is insufficient. Or if you think it would be a good idea to implement a generic solution for future controllers that would require more  complex gating.